# Database Admin 101 - Lab

## Introduction 

In this lab, you'll go through the process of designing and creating a database. From there, you'll begin to populate this table with mock data provided to you.

## Objectives

You will be able to:

* Use knowledge of the structure of databases to create a database and populate it

## The Scenario

You are looking to design a database for a school that will house various information from student grades to contact information, class roster lists and attendance. First, think of how you would design such a database. What tables would you include? What columns would each table have? What would be the primary means to join said tables?

## Creating the Database

Now that you've put a little thought into how you might design your database, it's time to go ahead and create it! Start by import the necessary packages. Then, create a database called **school.sqlite**.

In [27]:
import sqlite3
import pandas as pd

In [28]:
conn = sqlite3.connect('school.sqlite') #Creates the school.sqlite database

## Create a Table for Contact Information

Create a table called contactInfo to house contact information for both students and staff. Be sure to include columns for first name, last name, role (student/staff), telephone number, street, city, state, and zipcode. Be sure to also create a primary key for the table. 

In [29]:
cur = conn.cursor()
cur.execute("""CREATE TABLE contactinfo (
                                         userID INTEGER PRIMARY KEY,
                                         firstName TEXT,
                                         lastName TEXT,
                                         role TEXT,
                                         phoneNum INTEGER
                                         street TEXT
                                         city TEXT
                                         state TEXT
                                         zipCode TEXT
                                         );""")

OperationalError: table contactinfo already exists

## Populate the Table

Below, code is provided for you in order to load a list of dictionaries. Briefly examine the list. Each dictionary in the list will serve as an entry for your contact info table. Once you've briefly investigated the structure of this data, write a for loop to iterate through the list and create an entry in your table for each person's contact info.

In [30]:
# Code to load the list of dictionaries; just run this cell
import pickle

with open('contact_list.pickle', 'rb') as f:
    contacts = pickle.load(f)

print(contacts)

[{'firstName': 'Christine', 'lastName': 'Holden', 'role': 'staff', 'telephone ': 2035687697, 'street': '1672 Whitman Court', 'city': 'Stamford', 'state': 'CT', 'zipcode ': '06995'}, {'firstName': 'Christopher', 'lastName': 'Warren', 'role': 'student', 'telephone ': 2175150957, 'street': '1935 University Hill Road', 'city': 'Champaign', 'state': 'IL', 'zipcode ': '61938'}, {'firstName': 'Linda', 'lastName': 'Jacobson', 'role': 'staff', 'telephone ': 4049446441, 'street': '479 Musgrave Street', 'city': 'Atlanta', 'state': 'GA', 'zipcode ': '30303'}, {'firstName': 'Andrew', 'lastName': 'Stepp', 'role': 'student', 'telephone ': 7866419252, 'street': '2981 Lamberts Branch Road', 'city': 'Hialeah', 'state': 'Fl', 'zipcode ': '33012'}, {'firstName': 'Jane', 'lastName': 'Evans', 'role': 'student', 'telephone ': 3259909290, 'street': '1461 Briarhill Lane', 'city': 'Abilene', 'state': 'TX', 'zipcode ': '79602'}, {'firstName': 'Jane', 'lastName': 'Evans', 'role': 'student', 'telephone ': 32599092

In [31]:
# Your code to iterate over the contact list and populate the contactInfo table here
for contact in contacts:
    firstName = contact['firstName']
    lastName = contact['lastName']
    role = contact['role']
    telephone = contact['telephone ']
    street = contact['street']
    city = contact['city']
    state = contact['state']
    zipCode = contact['zipcode ']
    cur.execute("""INSERT INTO contactinfo (firstName, lastName, role, telephone, street, city, state, zipCode)
                  VALUES ('{}', '{}', '{}', '{}', '{}', '{}', '{}', '{}');""".format(firstName, lastName, role, telephone, street, city, state, zipCode) )
    

**Query the Table to Ensure it is populated**

In [32]:
cur.execute("""SELECT * FROM contactInfo;""")
df = pd.DataFrame(cur.fetchall())
df.columns = [x[0] for x in cur.description]
df

userId    firstName  lastName     role   telephone  \
0        1    Christine    Holden    staff  2035687697   
1        2  Christopher    Warren  student  2175150957   
2        3        Linda  Jacobson    staff  4049446441   
3        4       Andrew     Stepp  student  7866419252   
4        7         Mary    Raines  student  9075772295   
5        8           Ed     Lyman  student  5179695576   
6        9    Christine    Holden    staff  2035687697   
7       10  Christopher    Warren  student  2175150957   
8       11        Linda  Jacobson    staff  4049446441   
9       12       Andrew     Stepp  student  7866419252   
10      13         Jane     Evans  student  3259909290   
11      14         Jane     Evans  student  3259909290   
12      15         Mary    Raines  student  9075772295   
13      16           Ed     Lyman  student  5179695576   
14      17    Christine    Holden    staff  2035687697   
15      18  Christopher    Warren  student  2175150957   
16      19        Linda  Jacobson    staff  4049446441   
17      20       Andrew     Stepp  student  7866419252   
18      21         Jane     Evans  student  3259909290   
19      22         Jane     Evans  student  3259909290   
20      23         Mary    Raines  student  9075772295   
21      24           Ed     Lyman  student  5179695576   

                       street       city state zipcode  
0          1672 Whitman Court   Stamford    CT   06995  
1   1935 University Hill Road  Champaign    IL   61938  
2         479 Musgrave Street    Atlanta    GA   30303  
3   2981 Lamberts Branch Road    Hialeah    Fl   33012  
4       3975 Jerry Toth Drive  Ninilchik    AK   99639  
5         2910 Simpson Avenue       York    PA   17403  
6          1672 Whitman Court   Stamford    CT   06995  
7   1935 University Hill Road  Champaign    IL   61938  
8         479 Musgrave Street    Atlanta    GA   30303  
9   2981 Lamberts Branch Road    Hialeah    Fl   33012  
10        1461 Briarhill Lane    Abilene    TX   79602  
11        1461 Briarhill Lane    Abilene    TX   79602  
12      3975 Jerry Toth Drive  Ninilchik    AK   99639  
13              3478 Be Sreet    Lansing    MI   48933  
14         1672 Whitman Court   Stamford    CT   06995  
15  1935 University Hill Road  Champaign    IL   61938  
16        479 Musgrave Street    Atlanta    GA   30303  
17  2981 Lamberts Branch Road    Hialeah    Fl   33012  
18        1461 Briarhill Lane    Abilene    TX   79602  
19        1461 Briarhill Lane    Abilene    TX   79602  
20      3975 Jerry Toth Drive  Ninilchik    AK   99639  
21              3478 Be Sreet    Lansing    MI   48933

## Commit Your Changes to the Database

Persist your changes by committing them to the database.

In [33]:
conn.commit()

## Create a Table for Student Grades

Create a new table in the database called "grades". In the table, include the following fields: userId, courseId, grade.

** This problem is a bit more tricky and will require a dual key. (A nuance you have yet to see.)
Here's how to do that:

```SQL
CREATE TABLE table_name(
   column_1 INTEGER NOT NULL,
   column_2 INTEGER NOT NULL,
   ...
   PRIMARY KEY(column_1,column_2,...)
);
```

In [34]:
cur.execute("""CREATE TABLE grades(
                   userId INTEGER NOT NULL,
                   courseId INTEGER NOT NULL,
                   grade INTEGER
                   PRIMARY KEY(userID, courseID)
                   );
            """)

OperationalError: table grades already exists

## Remove Duplicate Entries

An analyst just realized that there is a duplicate entry in the contactInfo table! Find and remove it.

In [35]:
#find the duplicate entry
cur.execute("""SELECT firstName, lastName, telephone, COUNT(*)
               FROM contactInfo
               GROUP BY 1,2,3
               HAVING COUNT(*) > 1;""").fetchall()

[('Andrew', 'Stepp', 7866419252, 3),
 ('Christine', 'Holden', 2035687697, 3),
 ('Christopher', 'Warren', 2175150957, 3),
 ('Ed', 'Lyman', 5179695576, 3),
 ('Jane', 'Evans', 3259909290, 4),
 ('Linda', 'Jacobson', 4049446441, 3),
 ('Mary', 'Raines', 9075772295, 3)]

In [36]:
#delete the duplicate entry
cur.execute("""DELETE FROM contactInfo WHERE telephone = 3259909290""")

In [37]:
#check that your deleted entry was indeed removed
cur.execute("""SELECT firstName, lastName, telephone, COUNT(*)
               FROM contactInfo
               GROUP BY 1,2,3
               HAVING COUNT(*) > 1;""").fetchall()

[('Andrew', 'Stepp', 7866419252, 3),
 ('Christine', 'Holden', 2035687697, 3),
 ('Christopher', 'Warren', 2175150957, 3),
 ('Ed', 'Lyman', 5179695576, 3),
 ('Linda', 'Jacobson', 4049446441, 3),
 ('Mary', 'Raines', 9075772295, 3)]

## Updating an Address

Ed Lyman just moved to `2910 Simpson Avenue York, PA 17403`. Update his address accordingly.

In [45]:
#update Ed's address
cur.execute("""UPDATE contactInfo
               SET street = '2910 Simpson Avenue',
               SET city = 'York',
               SET state = 'PA'

SyntaxError: EOF while scanning triple-quoted string literal (<ipython-input-45-9ffae15f878e>, line 5)

In [24]:
#Your code here; Query the database to ensure the change was made

## Commit Your Changes to the Database

Once again, persist your changes by committing them to the database.

In [46]:
conn.commit()

## Summary

While there's certainly more to do with setting up and managing this database, you got a taste for creating, populating, and maintaining databases! Feel free to continue fleshing out this exercise for more practice. 